pkg/controller/discovery.go

```diff
@@ -0,0 +1,41 @@
+package controller
+
+import (
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+
+	"github.com/submariner-io/coastguard/pkg/remotecluster"
+)
+
+func (c *CoastguardController) OnAdd(clusterID string, kubeConfig *rest.Config) {
+	klog.Infof("adding cluster: %s", clusterID)
+
+	clientSet, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		klog.Errorf("error creating clientset for cluster %s: %s", clusterID, err.Error())
+		return
+	}
+
+	c.addCluster(clusterID, clientSet)
+}
+
+func (c *CoastguardController) addCluster(clusterID string, clientSet kubernetes.Interface) {
+	rc := remotecluster.New(clusterID, clientSet)
+	c.processingMutex.Lock()
+	c.remoteClusters[clusterID] = rc
+	c.processingMutex.Unlock()
```

Comment:
why not defer the unlock?

Reply:
So line 28 isn't executed holding the lock.

Reply:
instead of being smarty panties you could see that he is again asking for the lock there. So why not just remove that locking there and rely on the one out here...

Reply:
geez chill out man 😄 yes he does lock again in `onClusterFinishedSyncing` but that is called asynchronously from `rc.Run`.